### DISCUSSION TITLE
Proposal:  Allow null pattern to introduce variable designation

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#7404] [Language Ideas]
I'd like to propose that `null` patterns be extended to support a designation for a variable assignment which would allow more intuitive use of that pattern in some common cases.  The type of the variable would be the same as the expression being matched.

For example:

```cs
if (Some.Expression is null variable) {
    variable = SomeDefaultValue;
}

// use variable here
```

I'd also like to suggest that, unlike with normal type patterns where if they don't match the variable remains unassigned, that the variable is always considered definitely assigned to the value of the expression.  That would enable scenarios where one would prefer to use a `not` pattern to negate a `null` pattern but be able to extract the result into a variable:

```cs
if (Some.Expression is not null variable) {
    // use variable here
}
```

### Drawbacks

1. If the `null` pattern introduces a variable of the same type as the expression, that may create a situation with NVTs where the expected type of the variable would be non-nullable if the pattern is negated.  But I don't believe that this would be possible without the `not` pattern affecting the behavior of the `null` pattern and while keeping the pattern both in scope and definitely assigned when the pattern is or is not matched.
    ```cs
    int? nullableInt = Some.Expression;
    if (nullableInt is not null variable) {
        // expect variable to be int here?
    }
    ```

### Alternatives

1. Conjunctive patterns can be used to check and assign in a single expression:
    ```cs
    if (Some.Expression is not null and is var variable) {
        // use variable here
    }
    ```
3. The expression can be assigned to a variable in advance:
    ```cs
    T? variable = Some.Expression;
    if (variable is not null) {
        // use variable here
    }
    ```

**Comment by jnm2** [comment]
Previously asked at https://github.com/dotnet/csharplang/discussions/4071 and I'm pretty sure one other GitHub discussion where I asked @CyrusNajmabadi if he still wanted to do this (we had chatted about it long ago).

**Comment by bernd5** [comment]
We should allow that not only for `null` pattern. It should work for all constant patterns (numbers and strings).

**Reply by RenderMichael** [reply]
Extracting a variable from a constant would function similarly to how `not` works today, where the variable would only be relevant in the `else` block. Then `not const` would allow the extracted variable to be used then and there, much like how `not null` is used by OP

**Reply by bernd5** [reply]
In my eyes it would make sense even in the non-else case to avoid repeating...

**Reply by RenderMichael** [reply]
There's been some codegen issues with using a variable when the value is known, I've seen PRs into the runtime that basically change
```csharp
if (myBool == false)
{
    MyVar = myBool;
}
```
to 
```csharp
if (myBool == false)
{
    MyVar = false;
}
```
This could probably be improved but not without some work. Maybe Roslyn can even detect this. But it's a limitation of the runtime for now.
